idaes/apps/grid_integration/forecaster.py

```diff
@@ -251,3 +299,412 @@ def _forecast(self, means, stds, hour, horizon, n_samples):
+                )
+
+            if n_prices % 24 != 0:
+                raise ValueError(
+                    f"The number of prices for each bus should be a multiple of 24. But for bus {b}, {n_prices} are provided."
+                )
+
+            while len(historical_price[b]) // 24 > self.max_historical_days:
+                raise Warning(
+                    f"The number of days of the provided historical prices for bus {b} is greater than the max value {self.max_historical_days}. Dropping the first day's data."
+                )
+                historical_price[b] = historical_price[b][24:]
+
+    @property
+    def max_historical_days(self):
+
+        """
+        Property getter for max_historical_days.
+
+        Returns:
+            int: max historical days
+        """
+
+        return self._max_historical_days
+
+    @max_historical_days.setter
+    def max_historical_days(self, value):
+
+        """
+        Property setter for max_historical_days (validate before setting).
+
+        Args:
+            value: intended value for max_historical_days
+
+        Returns:
+            None
+        """
+
+        if not isinstance(value, Real):
+            raise TypeError(
+                f"max_historical_days must be a number, but {type(value)} is provided."
+            )
+
+        value = int(value)
+
+        if value < 1:
+            raise ValueError(
+                f"max_historical_days must be >= 1, but {value} is provided."
+            )
+
+        self._max_historical_days = value
+
+    @property
+    def historical_da_prices(self):
+
+        """
+        Property getter for historical_da_prices.
+
+        Returns:
+            dict: saved historical day-ahead prices
+        """
+
+        return self._historical_da_prices
+
+    @historical_da_prices.setter
+    def historical_da_prices(self, value):
+
+        """
+        Property setter for historical_da_prices (validate before setting).
+
+        Args:
+            value: intended value for historical_da_prices
+
+        Returns:
+            None
+        """
+
+        self._validate_input_historical_price(value)
+        self._historical_da_prices = value
+
+    @property
+    def historical_rt_prices(self):
+
+        """
+        Property getter for historical_rt_prices.
+
+        Returns:
+            dict: saved historical real-time prices
+        """
+
+        return self._historical_rt_prices
+
+    @historical_rt_prices.setter
+    def historical_rt_prices(self, value):
+
+        """
+        Property setter for historical_rt_prices (validate before setting).
+
+        Args:
+            value: intended value for historical_rt_prices
+
+        Returns:
+            None
+        """
+
+        self._validate_input_historical_price(value)
+        self._historical_rt_prices = value
+
+    def forecast_day_ahead_and_real_time_prices(
+        self, date, hour, bus, horizon, n_samples
```

Comment:
I think we need a test for this method. A single regression test would be fine.

Reply:
I have a question about this function. We have different forecast horizons for the day-ahead market and real-time market. DA is 48 and RT is 4. But in this function, we only pass one argument 'horizon' to the function. Will this lead to any problems?

Reply:
It looks like the day ahead bid calls this function with the day ahead horizon: https://github.com/xiangao1/idaes-pse/blob/65df44c044e8face10bff0c1613db91c2efe133b/idaes/apps/grid_integration/bidder.py#L538

But the real time bid calls the `forecast_real_time_prices` function with the real time horizon: https://github.com/xiangao1/idaes-pse/blob/65df44c044e8face10bff0c1613db91c2efe133b/idaes/apps/grid_integration/bidder.py#L575

So that seems the day ahead bid is using both the day ahead and real time forecast with the correct horizon length